### PR TITLE
Wandb k-fold cross-validation and hyperparameter search implementation

### PR DESCRIPTION
**Pull request by joshuakurien**

### CODE REVIEW

**Comment by joshuakurien** [comment]
Sample K-fold: https://wandb.ai/deep-breathe/ab_line/groups/kfold-f42alqy4/workspace?workspace=user-joshuakurien

File: config.yml [hunk 1]
```diff
@@ -43,12 +43,14 @@ WANDB:
   IMAGES_ARTIFACT_VERSION: ''
   MODEL_DEV_ARTIFACT_VERSION: ''
   TRAIN_VAL_TEST_ARTIFACT_VERSION: ''
+  K_FOLD_CROSS_VAL_ARTIFACT_VERSION: ''
   ARTIFACT_SEED: 42
 
 DATA:
   IMG_DIM: [128, 128]
   VAL_SPLIT: 0.1
```
Comment (bvanberl) [inline config.yml:51]: Did we end up deciding that this split would be a percentage of trainval or of the entire dataset? Just wanted to make sure that we don't have to update the documentation for this field in the readme.

**Review by bvanberl — APPROVED** [review]
Fantastic work!!

File: config.yml (hunk 1 above)
Comment (joshuakurien) [reply config.yml:51]: So VAL_SPLIT and TEST_SPLIT is for the entire dataset while K_FOLD_VALIDATION_SPLIT is just of trainval

Comment (bvanberl) [reply config.yml:51]: Cool. Could we document this field (and the others) in the readme? There's a section for the config fields.

File: src/train.py [hunk 2]
```diff
@@ -180,404 +163,176 @@ def on_epoch_end(self, epoch, logs=None):
...
+        )
+        wandb.agent(sweep_id, function=perform_single_run, count=cfg['TRAIN']['HPARAM_SEARCH']['N_EVALS'])
     elif experiment == 'cross_validation':
-        cross_validation(save_weights=save_weights, write_logs=write_logs)
+        # creates group id to be used for runs within the cross-fold validation iteration
+        val_group_id = f'kfold-{wandb.util.generate_id()}'
+        n_folds = get_n_folds(project_name=cfg['WANDB']['PROJECT_NAME'], entity_name=cfg['WANDB']['ENTITY'],
+                              artifact_version=cfg['WANDB']['K_FOLD_CROSS_VAL_ARTIFACT_VERSION'])
+        for fold_id in range(n_folds):
+            perform_single_run(group_id=val_group_id, fold_id=fold_id)
```
Comment (sdhossain) [inline src/train.py:330]: I take it this is not condensed into a sweep? (might be mistaken, please correct me if I'm wrong), but doing so might make it easier to group.

File: src/train.py [hunk 3]
```diff
@@ -180,404 +163,176 @@ def on_epoch_end(self, epoch, logs=None):
...
+            elif experiment_sweep_config[hparam_name]['TYPE'] == 'float_log':
+                parameter_config['distribution'] = 'log_uniform_values'
+                parameter_config['min'] = experiment_sweep_config[hparam_name]['RANGE'][0]
+                parameter_config['max'] = experiment_sweep_config[hparam_name]['RANGE'][1]
+            elif experiment_sweep_config[hparam_name]['TYPE'] == 'float_uniform':
+                parameter_config['distribution'] = 'uniform'
+                parameter_config['min'] = experiment_sweep_config[hparam_name]['RANGE'][0]
+                parameter_config['max'] = experiment_sweep_config[hparam_name]['RANGE'][1]
+            wandb_sweep_cfg['parameters'][hparam_name] = parameter_config
+
```
Comment (sdhossain) [inline src/train.py:291]: Really nice setup for configuring the parameters!

File: src/train.py (hunk 2 above)
Comment (sdhossain) [reply src/train.py:330]: perhaps you could even use the configure_hyperparameter_sweep and pass in a list like `{'fold_id': list(range(10))}` as the `experiment_sweep_config` argument

**Review by sdhossain — APPROVED** [review]
Looks great!

**Review by bvanberl — APPROVED** [review]

File: src/train_utils.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+    :param run: wandb run
+    :param artifact_version: version of artifact stored in wandb
+    :return: (Training DataFrame, validation DataFrame, test DataFrame, directory str for frames)
+    """
+
+    # uses the latest version of Images artifact if no version is specified
+    train_val_test_version = artifact_version if artifact_version else 'latest'
+
+    # downloads previously logged artifacts
+    train_val_test_artifact = run.use_artifact(f'TrainValTest:{train_val_test_version}')
```
Comment (sdhossain) [inline src/train_utils.py:33]: Do we want to make the artifacts configure-able as well? I know chances are one train-val-test is only going to be used, but depending on if we want to use completely different datasets (perhaps external).

File: src/train_utils.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        if cur_fold_id == fold_id:
+            test_df = pd.read_csv(fold_frames_path)
+        else:
+            train_val_dfs.append(pd.read_csv(fold_frames_path))
+    train_val_df = pd.concat(train_val_dfs)
+
+    train_df, val_df = group_train_test_split(train_val_df, test_size=val_split, group_key='patient_id',
+                                              target_key='Class', random_seed=random_seed)
+
+    return train_df, val_df, test_df, frames_dir
```
Comment (sdhossain) [inline src/train_utils.py:235]: Nicely done!

File: src/train_utils.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+# Evaluation callback generates prediction visualizations at the end of each epoch
+class WandbGradcamEvalCallback(WandbEvalCallback):
```
Comment (sdhossain) [inline src/train_utils.py:238]: Is GradCAM implementation included in here? (Wasn't able to spot it)

File: src/train.py [hunk 7]
```diff
@@ -180,404 +163,213 @@ def on_epoch_end(self, epoch, logs=None):
...
+            }
+        }
+    }
+
+    # Initialize sweep
+    sweep_id = wandb.sweep(
+        project=project_name,
+        entity=entity_name,
+        sweep=wandb_sweep_cfg
+    )
```
Comment (sdhossain) [inline src/train.py:339]: Nice implementation!

**Review by sdhossain — APPROVED** [review]
Overall very nicely done! I think it might be interesting to make artifact names also configure-able (perhaps for security considerations in the future where we don't want to expose wandb artifact name/info) or use different artifacts entirely (structured in the same manner).

We might need to document the necessary structure needed to make an artifact. (I believe this may have been automated in artifact creation, etc.)

File: src/train_utils.py (hunk 6 above)
Comment (joshuakurien) [reply src/train_utils.py:238]: No, it was included in the following PR.

File: src/train_utils.py (hunk 4 above)
Comment (joshuakurien) [reply src/train_utils.py:33]: This might be a future change we make in the Deep Breathe utils package.

**Pull request merged**